wazp/utils.py

```diff
@@ -142,110 +145,340 @@ def export_selected_rows_as_yaml(
     return
 
 
+def read_and_restructure_DLC_dataframe(
+    h5_file: str,
+) -> pd.DataFrame:
+    """Read and reorganise columns in DLC dataframe
+    The columns in the DLC dataframe as read from the h5 file are
+    reorganised to more closely match a long format.
+
+    The original columns in the DLC dataframe are multi-level, with
+    the following levels:
+    - scorer: if using the output from a model, this would be the model_str
+      (e.g. 'DLC_resnet50_jwasp_femaleandmaleSep12shuffle1_1000000').
+    - bodyparts: the keypoints tracked in the animal (e.g., head, thorax)
+    - coords: x, y, likelihood
+    # TODO: is this different in multianimal? (maybe an extra ID column?)
+
+    We reshape the dataframe to have a single level along the columns,
+    and the following columns:
+    - model_str: string that characterises the model used
+      (e.g. 'DLC_resnet50_jwasp_femaleandmaleSep12shuffle1_1000000')
+    - frame: the (zero-indexed) frame number the data was tracked at.
+      This is inherited from the index of the DLC dataframe.
+    - bodypart: the keypoints tracked in the animal (e.g., head, thorax).
+      Note we use the singular, rather than the plural as in DLC.
+    - x: x-coordinate of the bodypart tracked.
+    - y: y-coordinate of the bodypart tracked.
+    - likelihood: likelihood of the estimation provided by the model
+    The data is sorted by bodypart, and then by frame.
+
+    Parameters
+    ----------
+    h5_file : str
+        path to the input h5 file
+    Returns
+    -------
+    pd.DataFrame
+        a dataframe with the h5 file data, and the columns as specified above
+    """
+    # TODO: can this be less hardcoded?
+    # TODO: check with multianimal dataset!
+
+    # read h5 file as a dataframe
+    df = pd.read_hdf(h5_file)
+
+    # determine if model is multianimal
+    is_multianimal = "individuals" in df.columns.names
+
+    # assuming the DLC index corresponds to frame number!!!
+    # TODO: can I check this?
+    # frames are zero-indexed
+    df.index.name = "frame"
+
+    # stack scorer and bodyparts levels from columns to index
+    # if multianimal, also column 'individuals'
+    columns_to_stack = ["scorer", "bodyparts"]
+    if is_multianimal:
+        columns_to_stack.append("individual")
+    df = df.stack(level=columns_to_stack)  # type: ignore
```

Comment:
looks like [mypy has at least one issue](https://github.com/python/mypy/issues/3004) kindof related to this. (Pandas devs linked me to that).

I can also reproduce, and also don't understand your error message.
```
wazp/utils.py:204: error: Argument "level" to "stack" of "DataFrame" has incompatible type "List[str]"; expected "Union[Union[Hashable, int], List[Union[Hashable, int]]]"  [arg-type]
```

I don't understand why `List[str]` is not `List[Union[Hashable, ...]]`. And looking at the source, it seems mypy is ... parsing the docstring?? to get the type signature. Because pandas [aren't using explicit type hints](https://github.com/pandas-dev/pandas/blob/v1.5.3/pandas/core/frame.py#L8745).

All of that to say, I think it's correct for us to force an ignore here. I actually tend to think this is a mypy bug.

Reply:
FWIW I would tend to rely on our unit-testing rather than jumping through mypy hoops.